### PR TITLE
Light client bugfixes and updates

### PR DESCRIPTION
**Pull request by zsfelfoldi**

### CODE REVIEW

**Comment by fjl** [comment]
Please use the full import path in commit prefixes. e.g. "p2p/discv5: ...".

**Comment by zsfelfoldi** [comment]
@fjl done

**Comment by karalabe** [comment]
Btw, I'd like to only have bootnodes hosted by the foundation. Not sure where these come from.

File: p2p/discv5/net.go [hunk 1]
```diff
@@ -580,6 +597,14 @@ loop:
 		case <-refreshDone:
 			debugLog("<-net.refreshDone")
 			refreshDone = nil
+			list := searchReqWhenRefreshDone
+			searchReqWhenRefreshDone = nil
+			go func() {
+				time.Sleep(time.Second)
```
Comment (fjl) [inline p2p/discv5/net.go]: Why is there a delay here? If the refresh needs to wait for something else it's better to synchronise with that explicitly.

File: p2p/discv5/table.go [hunk 2]
```diff
@@ -64,42 +65,45 @@ func newTable(ourID NodeID, ourAddr *net.UDPAddr) *Table {
...
 	}
-	prefix = (prefix&addrMask ^ xorMask) | (rnd & rndMask)
-	binary.BigEndian.PutUint64(target[0:8], prefix)
+	targetPrefix := prefix ^ randUint64n(ddist)
+
+	var target common.Hash
+	binary.BigEndian.PutUint64(target[0:8], targetPrefix)
 	rand.Read(target[8:])
-	return common.BytesToHash(target)
+	return target
```
Comment (fjl) [inline p2p/discv5/table.go]: Can you explain the changes to this function?

**Comment by zsfelfoldi** [comment]
@karalabe these nodes are hosted on an Azure server that Luca has set me up and it belongs to the foundation. After we implement the port sharing, these will no longer be needed, we can use the same set of bootnodes for v4 and v5.

File: p2p/discv5/table.go (hunk 2 above)
Comment (zsfelfoldi) [inline p2p/discv5/table.go]: I wanted to do a random refresh target selection that selects closer addresses with a higher probability to fill up all buckets and ensure a healthy Kademlia topology. First I wanted to do this by selecting a certain bucket to refresh, but I realized that this approach had a lot of problems. Now I'm selecting a random node from the table and search in a distance radius of a similar order of magnitude. I believe this approach is both more reliable and effective.

**Comment by fjl** [comment]
What is the issue with selecting a bucket? It used to be impossible to do a lookup in a certain bucket but v5 makes it possible again. The kademlia paper recommends that the target should be the least recently used bucket.                                                                                                                                      
     
 
  
 
  Original Message  
Show Details
                                                                                                                                            
 
@zsfelfoldi commented on this pull request.





In p2p/discv5/table.go:


>  	rand.Read(target[8:])
-	return common.BytesToHash(target)
+	return target

I wanted to do a random refresh target selection that selects closer addresses with a higher probability to fill up all buckets and ensure a healthy Kademlia topology. First I wanted to do this by selecting a certain bucket to refresh, but I realized that this approach had a lot of problems. Now I'm selecting a random node from the table and search in a distance radius of a similar order of magnitude. I believe this approach is both more reliable and effective.

File: p2p/discv5/net.go (hunk 1 above)
Comment (zsfelfoldi) [inline p2p/discv5/net.go]: It was a test, I accidentally left it there, I'll remove it.

**Comment by zsfelfoldi** [comment]
@fjl I'd like to improve this selection later. What I'm doing currently may be not ideal but close to it. The previous code was buggy and I wanted something simple and reliable for now. If it's OK, I'd like to merge it like this for the first release and then rewrite it later (maybe fix the original version), along with a lot of other things that need improvement in this code.

**Comment by fjl** [comment]
I'm asking like this because I find the change hard to understand and would prefer a simplified change with less magic. In v4 the refresh logic is to perform a random lookup. This is sufficient most of the time. 

I'm willing to accept almost any code in p2p/discv5 because v5 is new and experimental. It would still be nice to if you added a comment that explains what the target algorithm is trying to do because that's not visible at all.

**Comment by zsfelfoldi** [comment]
I added a comment. It wouldn't be hard to do the final algorithm now but I don't want to risk breaking it again so close to the release.

**Pull request merged**